src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Regex.EnumerateMatches.cs

```diff
@@ -0,0 +1,144 @@
+        /// <see cref="MoveNext"/> returns <see langword="true"/> and <see cref="Current"/> will contain the first <see cref="ValueMatch"/>. If there are no successful matches,
+        /// then <see cref="MoveNext"/> returns <see langword="false"/> and <see cref="Current"/> throws an <see cref="InvalidOperationException"/>.
+        ///
+        /// This type is a ref struct since it stores the input span as a field in order to be able to lazily iterate over it.
+        /// </remarks>
+        public ref struct ValueMatchEnumerator
+        {
+            private readonly Regex _regex;
+            private readonly ReadOnlySpan<char> _input;
+            private ValueMatch _matchReference;
+            private int _startAt;
+            private int _prevLen;
+            private bool _hasValidValue;
+
+            /// <summary>
+            /// Creates an instance of the <see cref="ValueMatchEnumerator"/> for the passed in <paramref name="regex"/> which iterates over <paramref name="input"/>.
+            /// </summary>
+            /// <param name="regex">The <see cref="Regex"/> to use for finding matches.</param>
+            /// <param name="input">The input span to iterate over.</param>
+            /// <param name="startAt">The position where the engine should start looking for matches from.</param>
+            internal ValueMatchEnumerator(Regex regex, ReadOnlySpan<char> input, int startAt)
+            {
+                _regex = regex;
+                _input = input;
+                _matchReference = default;
+                _startAt = startAt;
+                _prevLen = -1;
+                _hasValidValue = false;
+            }
+
+            /// <summary>
+            /// Provides an enumerator that iterates through the matches in the input span.
+            /// </summary>
+            /// <returns>A copy of this enumerator.</returns>
+            public readonly ValueMatchEnumerator GetEnumerator() => this;
+
+            /// <summary>
+            /// Advances the enumerator to the next match in the span.
+            /// </summary>
+            /// <returns>
+            /// <see langword="true"/> if the enumerator was successfully advanced to the next element; <see langword="false"/> if the enumerator cannot find additional matches.
+            /// </returns>
+            public bool MoveNext()
+            {
+                Match? match = _regex.RunSingleMatch(quick: false, _prevLen, _input, _startAt);
+                if (match is not null && match != RegularExpressions.Match.Empty)
+                {
+                    _matchReference = new ValueMatch(match);
+                    _hasValidValue = true;
+                    _startAt = match._textpos;
+                    _prevLen = match.Length;
+                    return true;
+                }
+                _hasValidValue = false;
+                _matchReference = default;
```

Comment:
This shouldn't be needed either. See what the C# compiler generates for an enumerator:
https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA+ABATARgLABQGADAAQY4oDchGAzOVqQMKkDehpX5DFdAPAEsAdgBcAfKQCyACgCU7Tt2UUc5AOykUWGgWUBfQvqA===